### DISCUSSION TITLE
Configuring mitmproxy (reverse proxy) as basic WAF - is it possible?

### DISCUSSION THREAD

**Discussion by dorsis** [mitmproxy/mitmproxy#5306] [Q&A]
Hi all,
I'm looking for a way to configure mitmproxy (reverse proxy) as a basic WAF, that will intercept client requests (before sent to server), and server response (before sent back to client).
I came up with a suggested flow, see attached.
Would be happy to hear your inputs on this... Is it possible with mitmproxy custom scripting capabilities?
[architecture diagram - mitm.pdf](https://github.com/mitmproxy/mitmproxy/files/8590380/architecture.diagram.-.mitm.pdf)

Thanks !

**Comment by mhils** [comment]
That doesn't look particularly performant, but if that's fine for your use case that's okay. I think you can do requests to your service using an async HTTP request library like [httpx](https://www.python-httpx.org/) in the [`request`/`response`](https://docs.mitmproxy.org/dev/api/events.html#HTTPEvents) hooks and call `flow.kill()` (or set/override `flow.response`) when you want to block.

Here's a simple (untested) sketch:


```python
import httpx
from mitmproxy import http

async def request(flow: http.HTTPFlow):
    async with httpx.AsyncClient() as client:
        r = await client.get('https://www.example.com/')  # change this to include the relevant parts from the request
        if r.status_code == 418:
            flow.kill()  # or set flow.response, see the `http-reply-from-proxy.py` example.
```
